packages/belos/tpetra/src/solvers/Belos_Tpetra_GmresPipeline.hpp

```diff
@@ -293,34 +310,64 @@ class GmresPipeline : public Gmres<SC, MV, OP> {
         y.resize (restart+1);
       }
       // Compute real residual
-      Z = * (V.getVectorNonConst (0));
-      A.apply (X, Z);
-      Z.update (one, B, -one);
-      r_norm = Z.norm2 (); // residual norm
+      A.apply (X, R);
+      R.update (one, B, -one);
+      // TODO: compute residual norm with all-reduce, should be idot?
+      r_norm = R.norm2 ();
```

Comment:
See above.  It may be just as important here to have a fused residual kernel here, as to exploit overlap.